File: ios/Utils/CloudStorageError.swift
//
//  CloudStorageError.swift
//  CloudStorage
//
//  Created by Maximilian Krause on 27.09.24.
//  Copyright © 2024 Kuatsu App Agency. All rights reserved.
//

import Foundation

enum CloudStorageError: Error {
    case invalidScope(scope: String)
    case fileNotFound(path: String)
    case pathIsDirectory(path: String)
    case pathIsFile(path: String)
    case directoryNotFound(path: String)
    case directoryNotEmpty(path: String)
    case fileAlreadyExists(path: String)
    case authenticationFailed
    case writeError(path: String)
    case readError(path: String)
    case deleteError(path: String)
    case statError(path: String)
    case unknown(message: String = "An unknown error occurred")
    case fileNotDownloadable(path: String)

    var code: String {
        switch self {
        case .invalidScope: return "ERR_INVALID_SCOPE"
        case .fileNotFound: return "ERR_FILE_NOT_FOUND"
        case .pathIsDirectory: return "ERR_PATH_IS_DIRECTORY"
        case .pathIsFile: return "ERR_PATH_IS_FILE"
        case .directoryNotFound: return "ERR_DIRECTORY_NOT_FOUND"
        case .directoryNotEmpty: return "ERR_DIRECTORY_NOT_EMPTY"
        case .fileAlreadyExists: return "ERR_FILE_EXISTS"
        case .authenticationFailed: return "ERR_AUTHENTICATION_FAILED"
        case .writeError: return "ERR_WRITE_ERROR"
        case .readError: return "ERR_READ_ERROR"
        case .deleteError: return "ERR_DELETE_ERROR"
        case .statError: return "ERR_STAT_ERROR"
        case .unknown: return "ERR_UNKNOWN"
        case .fileNotDownloadable: return "ERR_FILE_NOT_DOWNLOADABLE"
        }
    }

    var message: String {
        switch self {
        case .invalidScope(let scope):
            return "Invalid scope \(scope) provided"
        case .fileNotFound(let path):
            return "File not found at path \(path)"
        case .pathIsDirectory(let path):
            return "Path is a directory at path \(path)"
        case .pathIsFile(let path):
            return "Path is a file at path \(path)"
        case .directoryNotFound(let path):
            return "Directory not found at path \(path)"
        case .directoryNotEmpty(let path):
            return "Directory not empty at path \(path)"
        case .fileAlreadyExists(let path):
            return "File already exists at path \(path)"
        case .authenticationFailed:
            return "Authentication failed"
        case .writeError(let path):
            return "Write error for path \(path)"
        case .readError(let path):
            return "Read error for path \(path)"
        case .deleteError(let path):
            return "Delete error for path \(path)"
        case .statError(let path):
            return "Stat error for path \(path)"
        case .unknown(let message):
            return message
        case .fileNotDownloadable(let path):
            return "File not downloadable at path \(path)"
        }
    }
}

File: packages/react-native-cloud-storage/android/src/main/java/com/voicekit/CloudStorageError.kt
package com.cloudstorage

sealed class CloudStorageError(
    val code: String,
    val message: String
) : Exception(message) {
    data class InvalidScope(val scope: String) : CloudStorageError(
        code = "ERR_INVALID_SCOPE",
        message = "Invalid scope $scope provided"
    )

    data class FileNotFound(val path: String) : CloudStorageError(
        code = "ERR_FILE_NOT_FOUND",
        message = "File not found at path $path"
    )

    data class PathIsDirectory(val path: String) : CloudStorageError(
        code = "ERR_PATH_IS_DIRECTORY",
        message = "Path is a directory at path $path"
    )

    data class PathIsFile(val path: String) : CloudStorageError(
        code = "ERR_PATH_IS_FILE",
        message = "Path is a file at path $path"
    )

    data class DirectoryNotFound(val path: String) : CloudStorageError(
        code = "ERR_DIRECTORY_NOT_FOUND",
        message = "Directory not found at path $path"
    )

    data class DirectoryNotEmpty(val path: String) : CloudStorageError(
        code = "ERR_DIRECTORY_NOT_EMPTY",
        message = "Directory not empty at path $path"
    )

    data class FileAlreadyExists(val path: String) : CloudStorageError(
        code = "ERR_FILE_EXISTS",
        message = "File already exists at path $path"
    )

    object AuthenticationFailed : CloudStorageError(
        code = "ERR_AUTHENTICATION_FAILED",
        message = "Authentication failed"
    )

    data class WriteError(val path: String) : CloudStorageError(
        code = "ERR_WRITE_ERROR",
        message = "Write error for path $path"
    )

    data class ReadError(val path: String) : CloudStorageError(
        code = "ERR_READ_ERROR",
        message = "Read error for path $path"
    )

    data class DeleteError(val path: String) : CloudStorageError(
        code = "ERR_DELETE_ERROR",
        message = "Delete error for path $path"
    )

    data class StatError(val path: String) : CloudStorageError(
        code = "ERR_STAT_ERROR",
        message = "Stat error for path $path"
    )

    data class FileNotDownloadable(val path: String) : CloudStorageError(
        code = "ERR_FILE_NOT_DOWNLOADABLE",
        message = "File not downloadable at path $path"
    )

    data class Unknown(val errorMessage: String = "An unknown error occurred") : CloudStorageError(
        code = "ERR_UNKNOWN",
        message = errorMessage
    )
}
